src/v/cluster/node_status_provider.cc

```diff
@@ -0,0 +1,235 @@
+/*
+ * Copyright 2022 Redpanda Data, Inc.
+ *
+ * Use of this software is governed by the Business Source License
+ * included in the file licenses/BSL.md
+ *
+ * As of the Change Date specified in that file, in accordance with
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
+
+#include "cluster/node_status_provider.h"
+
+#include "cluster/cluster_utils.h"
+#include "cluster/errc.h"
+#include "cluster/logger.h"
+#include "config/node_config.h"
+#include "ssx/future-util.h"
+
+namespace cluster {
+
+node_status_provider::node_status_provider(
+  model::node_id self,
+  members_table& members_table,
+  config::binding<std::chrono::milliseconds> period)
+  : _self(self)
+  , _members_table(members_table)
+  , _period(std::move(period))
+  , _rpc_tls_config(config::node().rpc_server_tls()) {
+    if (!config::shard_local_cfg().disable_public_metrics()) {
+        setup_metrics(_public_metrics);
+    }
+
+    if (!config::shard_local_cfg().disable_metrics()) {
+        setup_metrics(_metrics);
+    }
+
+    _members_table_notification_handle
+      = _members_table.register_members_updated_notification(std::bind_front(
+        std::mem_fn(&node_status_provider::handle_members_updated_notification),
+        this));
+
+    _timer.set_callback([this] { tick(); });
+}
+
+void node_status_provider::handle_members_updated_notification(
+  std::vector<model::node_id> node_ids) {
+    for (const auto& node_id : node_ids) {
+        if (node_id == _self || _discovered_peers.contains(node_id)) {
+            continue;
+        }
+
+        vlog(
+          clusterlog.info,
+          "Node {} has been discovered via members table",
+          node_id);
+        _discovered_peers.insert(node_id);
+    }
+}
+
+void node_status_provider::start() {
+    vassert(ss::this_shard_id() == shard, "invoked on a wrong shard");
+    return tick();
+}
+
+ss::future<> node_status_provider::stop() {
+    vassert(ss::this_shard_id() == shard, "invoked on a wrong shard");
+
+    _metrics.clear();
+    _public_metrics.clear();
+
+    _members_table.unregister_members_updated_notification(
+      _members_table_notification_handle);
+    _timer.cancel();
+    return _gate.close().then([this] { return _node_connection_cache.stop(); });
+}
+
+std::optional<node_status>
+node_status_provider::get_node_status(model::node_id node) const {
+    if (node == _self) {
+        return node_status{
+          .node_id = node, .last_seen = rpc::clock_type::now()};
+    }
+
+    if (_peers_state.contains(node)) {
```

Comment:
nit: you can spare one lookup when using `find`

Reply:
Done.